### PR TITLE
fix: handle route encoding

### PR DESCRIPTION
**Pull request by pi0**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

vuejs/vue-router#3350 introduced a breaking change which dropped supporting to encode uri paths. So we need to update internals doing it:

- Ensure `router.base` is always encoded
- Encode `path` param in route entries (generated code) (update #8421)
- Decode URIs for CLI messages for DX and avoid conflicting with util.format
- Override `router.resolve` with `encodeURI(decodeURI(to))` (props to @farnabaz)
- Update rest of internals
- New encoding test (local): `yarn nuxt dev test/fixtures/encoding`
  - Would be nice if we have more e2e tests later on specially for SPA and CSR

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

**Review by clarkdo — DISMISSED** [review]

**Comment by pi0** [comment]
merging to try on edge

File: yarn.lock [hunk 1]
```diff
@@ -13403,6 +13403,11 @@ vue-router@3.4.8:
   resolved "https://registry.npmjs.org/vue-router/-/vue-router-3.4.8.tgz#2c06261d35d8075893470352d42d70b6287b8194"
   integrity sha512-3BsR84AqarcmweXjItxw3jwQsiYNssYg090yi4rlzTnCJxmHtkyCvhNz9Z7qRSOkmiV485KkUCReTp5AjNY4wg==
 
+vue-router@^3.4.9:
+  version "3.4.9"
+  resolved "https://registry.npmjs.org/vue-router/-/vue-router-3.4.9.tgz#c016f42030ae2932f14e4748b39a1d9a0e250e66"
+  integrity sha512-CGAKWN44RqXW06oC+u4mPgHLQQi2t6vLD/JbGRDAXm0YpMv0bgpKuU5bBd7AvMgfTz9kXVRIWKHqRwGEb8xFkA==
+
```
Comment (danielroe) [inline yarn.lock:13410]: should replace router above

File: packages/vue-app/template/server.js [hunk 2]
```diff
@@ -7,7 +7,8 @@ import {
   <% if (features.middleware) { %>middlewareSeries,<% } %>
   <% if (features.middleware && features.layouts) { %>sanitizeComponent,<% } %>
   getMatchedComponents,
-  promisify
+  promisify,
+  ensureURIEncoded
```
Comment (danielroe) [inline packages/vue-app/template/server.js:11]:
```suggestion
  promisify
```

**Review by danielroe — APPROVED** [review]

File: yarn.lock (hunk 1 above)
Comment (pi0) [reply yarn.lock:13410]: (handled in renovate PR)

File: packages/vue-app/template/server.js (hunk 2 above)
Comment (pi0) [reply packages/vue-app/template/server.js:11]: Nice catch! a8cc430574ef02a7817da357fb9daf22de74778f

File: packages/server/src/server.js [hunk 3]
```diff
@@ -167,7 +167,7 @@ export default class Server {
         prefix: false,
         handler: (req, res) => {
           const to = urlJoin(this.nuxt.options.router.base, req.url)
-          consola.info(`[Development] Redirecting from \`${req.url}\` to \`${to}\` (router.base specified).`)
+          consola.info(`[Development] Redirecting from \`${decodeURI(req.url)}\` to \`${decodeURI(to)}\` (router.base specified).`)
```
Comment (danielroe) [inline packages/server/src/server.js:170]: I'm sure there's a reason - but is there any merit in normalising on the `req` object?

Comment (pi0) [reply packages/server/src/server.js:170]: Actually it is a bad-request error for any http server with non-encoded URI. It is mostly to ensure not happening if `req` is not real `req` :)

**Pull request merged**